### PR TITLE
Make Qulacs compile with Clang and as sub-project / bug fix for USE_PYTHON=ON

### PR DESCRIPTION
**Pull request by mmoelle1**
These minor changes fix the following issues:

- Qulacs cannot be used as CMake sub-project (replaced `CMAKE_[SOURCE,BINARY]_DIR` by `PROJECT_[SOURCE,BINARY]_DIR`
- Qulacs does not compile with Clang (replaced hard-coded `-fopenmp` flag by CMake's FindOpenMP mechanism)
- Qulacs does not compile with `USE_PYTHON=OFF` (fixed wrong dependency)

### CODE REVIEW

**Review by Hiroya-W — COMMENTED** [review]
Thank you for PR! Could you please check my comments?

File: CMakeLists.txt
Comment (mmoelle1) [reply CMakeLists.txt:271]: Fully agree!

File: CMakeLists.txt
Comment (mmoelle1) [reply CMakeLists.txt:518]: I think that the targets `qulacs_core` and `eigen` are not visible in `python/CMakeLists.txt` so the dependency needs to be handled in the main `./CMakeLists.txt` file. 

It might be more readable to ident the two lines which I did in my forked repository. Please let me know if I should change anything else in the PR.

Comment (Hiroya-W) [reply CMakeLists.txt:518]: > I think that the targets qulacs_core and eigen are not visible in python/CMakeLists.txt

The `qulacs_core` library is defined in this line of `python/CMakeLists.txt`.
https://github.com/qulacs/qulacs/blob/d6306cecc754f315e87cd4b2ee53d592528ae28c/python/CMakeLists.txt#L36

The `eigen` is defined in `./CMakeLists.txt`, but when `USE_PYTHON=Yes`, the subdirectory is loaded, so it is possible to find the dependency.

https://github.com/qulacs/qulacs/blob/d6306cecc754f315e87cd4b2ee53d592528ae28c/CMakeLists.txt#L415-L417

For example, `pybind11_pop` is added as a dependency in `python/CMakeLists.txt`.

https://github.com/qulacs/qulacs/blob/d6306cecc754f315e87cd4b2ee53d592528ae28c/python/CMakeLists.txt#L42

What do you think?

Comment (mmoelle1) [reply CMakeLists.txt:518]: I fully agree. How would you like me to change the PR? Shall I only move the `add_dependencies(qulacs_core eigen)` to `python/CMakeLists.txt` or everything that is needed for PyBind11 (which includes fetching the eigen library)?

Comment (Hiroya-W) [reply CMakeLists.txt:518]: Please can you move only `add_dependencies(qulacs_core eigen)` to `python/CMakeLists.txt`? I think that should work fine. Thank you!

Comment (mmoelle1) [reply CMakeLists.txt:515]: Done so. I you want me to make any additional changes before merging this PR please let me know.

Comment (ikanago) [reply CMakeLists.txt:515]: I also checked that moving `add_dependencies(qulacs_core eigen)` to `python/CMakeLists.txt` works well.

**Review by ikanago — APPROVED** [review]
Thank you for your contribution! I will merge this PR after I checked the CI passed.
There is no CI job to compile the project with clang. So I will set it up in another branch if needed.

**Pull request merged**
